README.md
Software Heritage - CVS loader
==============================

The Software Heritage CVS Loader imports the history of CVS repositories
into the SWH dataset.

The main entry points are

- :class:`swh.loader.cvs.loader.CvsLoader` for the main cvs loader which ingests content out of
  a local cvs repository


Features
--------
The CVS loader can access CVS repositories via rsync or via the CVS
pserver protocol, with optional support for tunnelling pserver via SSH.

The CVS loader does _not_ require the cvs program to be installed.

Access via rsync requires the rsync program to be installed.
The CVS loader will then invoke rsync to obtain a temporary local copy of the
entire CVS repository. It will then walk the local copy the CVS repository and
parse history of each RCS file with a built-in RCS parser.
This will usually be the fastest method for importing a given CVS repository.
However, most CVS servers do not offer repository access via rsync, and CVS
repositories which see active commits may see conversion problems because the
CVS repository format was not designed for lock-less read access.

Access via the plaintext CVS pserver protocol requires no external dependencies
to be installed, and is compatible with regular CVS servers. This method will
use read-locks on the server side and should therefore be safe to use with
active CVS repositories.
The CVS loader will use a built-in minimal CVS client written in Python to fetch
the output of the cvs rlog command executed on the CVS server. This output will
be processed to obtain repository history information. All versions of all files
will then be fetched from the server and injected into the SWH archive.

Access via pserver over SSH requires OpenSSH to be installed. Apart from
using SSH as a transport layer the conversion process is the same as in
the plaintext pserver case. The SSH client will be instructed to trust SSH host key
fingeprints upon first use. If a CVS server changes its SSH fingerprint then manual
intervention may be required in order for future visits to be successful.

Regardless of access protocol, the CVS loader uses heuristics to convert the
per-file history stored in CVS into changesets. These changesets correspond to
snapshots in the SWH database model. A given CVS repository should always yield
a consistent series of changesets across multiple visits.

The following URL protocol schemes are recognized by the loader:

- rsync://
- pserver://
- ssh://

After the protocol scheme, the CVS server hostname must be specified,
with an optional user:password field delimited from the hostname
with the '@' character:

```
pserver://anonymous:password@cvs.example.com/
```

After the hostname, the server-side CVS root path must be specified.
The path will usually contain a CVSROOT directory on the server, though
this directory may be hidden from clients:

```
pserver://anonymous:password@cvs.example.com/var/cvs/
```

The final component of the URL identifies the name of the CVS module
which should be ingested into the SWH archive:

```
pserver://anonymous:password@cvs.example.com/var/cvs/project1
```

As a concrete example, this URL points to the historical CVS repository
of the a2ps project. In this case, the cvsroot path is /sources/a2ps and
the CVS module of the project is called a2ps:

```
pserver://anonymous:anonymous@cvs.savannah.gnu.org/sources/a2ps/a2ps
```

In order to obtain the history of this repository the CVS loader will
perform the CVS pserver protocol exchange which is also performed by:

```
cvs -d :pserver:anonymous@cvs.savannah.gnu.org/sources/a2ps rlog a2ps
```

Known Limitations
-----------------
CVS repositories which see active commits should be converted with care.
It is possible to end up with a partial conversion of the latest commit
if repository data is fetched via rsync while a commit is in progress.
The pserver protocol is the safer option in such cases.

Only history of main CVS branch is converted.
CVS vendor branch imports and merges which modify the main branch are
modeled as two distinct commits to the main branch.
Other branches will not be represented in the conversion result at all.

CVS labels are not converted into corresponding SWH tags/releases yet.

The converter does not yet support incremental fetching of CVS history.
The entire history will be fetched and processed during every visit.
By design, CVS does not fully support a concept of changesets that span multiple
files and, as such, importing an evolving CVS history incrementally is a not a
trivial problem. Regardless, some improvements could be made relatively easily,
as noted below.

CVS repositories copied with rsync could be cached locally, such that
rsync will only download RCS files which have changed since the last visit.
At present the local copy of the repository is fetched to a temporary directory
and is deleted once the conversion process is done.

It might help to store persistent meta-data about blobs imported from CVS.
If such meta-data could be searched via a given CVS repository name, a path,
and an RCS revision number then redundant downloads of file versions over
the pserver protocol could be detected and skipped.

The minimal CVS client does not yet support the optional gzip extension
offered by the CVS pserver protocol. If this was supported then files
downloaded from a CVS server could be compressed while in transit.

The built-in minimal CVS client has not been tested against many versions of CVS.
It should work fine against CVS 1.11 and 1.12 servers. More work may be needed
to improve compatibility with older versions of CVS.

Acknowledgements
----------------
This software contains code derived from *cvs2gitdump* written by YASUOKA Masahiko,
and from the *rcsparse* library written by Simon Schubert.

This software contains code derived from ViewVC: https://www.viewvc.org/

Licensing information
---------------------
Parts of the software written by SWH developers are licensed under GPLv3.
See the file LICENSE

cvs2gitdump by YASUOKA Masahiko is licensed under ISC.
See the top of the file swh/loader/cvs/cvs2gitdump/cvs2gitdump.py

rcsparse by Simon Schubert is licensed under AGPLv3.
See the file swh/loader/cvs/rcsparse/COPYRIGHT

ViewVC is licensed under the 2-clause BSD licence.
See the file swh/loader/cvs/rlog.py

# Running Tests

Because the rcsparse library is implemented in C and accessed via Python bindings,
the CVS loader must be compiled and installed before tests can be run and the
*build* directory must be passed as an argument to pytest:

```
$ ./setup.py build install
$ pytest ./build
```

# CLI run

With the configuration:

/tmp/loader_cvs.yml:
```
storage:
  cls: remote
  args:
    url: http://localhost:5002/
```

Run:

```
swh loader --config-file /tmp/loader_cvs.yml \
    run cvs <cvs-url>
```


debian/changelog
swh-loader-cvs (0.7.0-1~swh1~bpo10+1) buster-swh; urgency=medium

  * Rebuild for buster-swh

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Wed, 28 Jun 2023 09:28:57 +0000

swh-loader-cvs (0.7.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.7.0     - (tagged by Antoine R. Dumont
    (@ardumont) <ardumont@softwareheritage.org> on 2023-06-28 11:21:03
    +0200)
  * Upstream changes:     - v0.7.0     - Ensure tasks can be scheduled
    - loader: Handle CVS module with no changesets     - loader: Raise
    NotFound when CVS repo no longer fetchable with rsync     - loader:
    Fix loading when CVSROOT/config file is missing

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Wed, 28 Jun 2023 09:27:11 +0000

swh-loader-cvs (0.6.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.6.0     - (tagged by Antoine R. Dumont
    (@ardumont) <ardumont@softwareheritage.org> on 2023-03-13 12:17:50
    +0100)
  * Upstream changes:     - v0.6.0     - cvsclient: Do not abort
    checkout when skipping $Log$ keyword expansion

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Mon, 13 Mar 2023 11:21:26 +0000

swh-loader-cvs (0.5.1-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.1     - (tagged by Antoine Lambert
    <anlambert@softwareheritage.org> on 2022-10-17 19:47:59 +0200)
  * Upstream changes:     - version 0.5.1

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Mon, 17 Oct 2022 17:51:56 +0000

swh-loader-cvs (0.5.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.0     - (tagged by Antoine Lambert
    <anlambert@softwareheritage.org> on 2022-10-17 19:26:13 +0200)
  * Upstream changes:     - version 0.5.0

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Mon, 17 Oct 2022 17:30:12 +0000

swh-loader-cvs (0.4.1-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.4.1     - (tagged by Antoine Lambert
    <anlambert@softwareheritage.org> on 2022-09-19 10:53:59 +0200)
  * Upstream changes:     - version 0.4.1

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Mon, 19 Sep 2022 08:58:08 +0000

swh-loader-cvs (0.4.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.4.0     - (tagged by Antoine Lambert
    <anlambert@softwareheritage.org> on 2022-09-15 17:05:42 +0200)
  * Upstream changes:     - version 0.4.0

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Thu, 15 Sep 2022 15:11:17 +0000

swh-loader-cvs (0.3.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.3.0     - (tagged by Antoine R. Dumont
    (@ardumont) <ardumont@softwareheritage.org> on 2022-05-11 09:07:04
    +0200)
  * Upstream changes:     - v0.3.0     - cvs.loader: Decrease log level
    verbosity to debug     - add strict asyncio_mode in pytest.ini     -
    add missing sentry captures

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Wed, 11 May 2022 07:10:34 +0000

swh-loader-cvs (0.2.2-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.2.2     - (tagged by Antoine Lambert
    <anlambert@softwareheritage.org> on 2022-04-29 12:10:37 +0200)
  * Upstream changes:     - version 0.2.2

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Fri, 29 Apr 2022 10:13:55 +0000

swh-loader-cvs (0.2.1-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.2.1     - (tagged by Antoine R. Dumont
    (@ardumont) <ardumont@softwareheritage.org> on 2022-04-25 18:09:06
    +0200)
  * Upstream changes:     - v0.2.1     - setup.py: Fix description
    content-type to markdown

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Mon, 25 Apr 2022 16:12:32 +0000

swh-loader-cvs (0.1.1-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.1.1     - (tagged by Antoine Lambert
    <anlambert@softwareheritage.org> on 2022-04-14 17:53:33 +0200)
  * Upstream changes:     - version 0.1.1

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Thu, 14 Apr 2022 15:58:21 +0000

swh-loader-cvs (0.1.0-1~swh2) unstable-swh; urgency=medium

  * Fix dependency and bump new release

 -- Antoine R. Dumont (@ardumont) <ardumont@softwareheritage.org>  Fri, 07 Jan 2022 15:30:29 +0100

swh-loader-cvs (0.1.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.1.0     - (tagged by Antoine R. Dumont
    (@ardumont) <ardumont@softwareheritage.org> on 2022-01-07 15:15:00
    +0100)
  * Upstream changes:     - v0.1.0     - Validate input paths in the CVS
    loader     - swh.loader.cvs.tasks: Fix parameter uses to the ones
    needed

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Fri, 07 Jan 2022 14:17:40 +0000

swh-loader-cvs (0.0.2-2~swh1) unstable-swh; urgency=medium

  * Add missing dh_install override to avoid stomping on the namespace
    __init__.py

 -- Nicolas Dandrimont <olasd@debian.org>  Wed, 15 Dec 2021 15:43:29 +0100

swh-loader-cvs (0.0.2-1~swh1) unstable-swh; urgency=medium

  * Initial release

 -- Nicolas Dandrimont <nicolas@dandrimont.eu>  Wed, 15 Dec 2021 10:50:25 +0100


swh/loader/cvs/cvsclient.py
# Copyright (C) 2015-2022  The Software Heritage developers
# See the AUTHORS file at the top-level directory of this distribution
# License: GNU Affero General Public License version 3, or any later version
# See top-level LICENSE file for more information

"""Minimal CVS client implementation

"""

import os.path
import socket
import subprocess
import tempfile
from typing import IO, Tuple

from tenacity import retry
from tenacity.retry import retry_if_exception_type
from tenacity.stop import stop_after_attempt

from swh.loader.exception import NotFound

CVS_PSERVER_PORT = 2401
CVS_PROTOCOL_BUFFER_SIZE = 8192
EXAMPLE_PSERVER_URL = "pserver://user:password@cvs.example.com/cvsroot/repository"
EXAMPLE_SSH_URL = "ssh://user@cvs.example.com/cvsroot/repository"

VALID_RESPONSES = [
    "ok",
    "error",
    "Valid-requests",
    "Checked-in",
    "New-entry",
    "Checksum",
    "Copy-file",
    "Updated",
    "Created",
    "Update-existing",
    "Merged",
    "Patched",
    "Rcs-diff",
    "Mode",
    "Removed",
    "Remove-entry",
    "Template",
    "Notified",
    "Module-expansion",
    "Wrapper-rcsOption",
    "M",
    "Mbinary",
    "E",
    "F",
    "MT",
]

# Trivially encode strings to protect them from innocent eyes (i.e.,
# inadvertent password compromises, like a network administrator
# who's watching packets for legitimate reasons and accidentally sees
# the password protocol go by).
#
# This is NOT secure encryption.


def scramble_password(password):
    s = ["A"]  # scramble scheme version number
    # fmt: off
    scramble_shifts = [
        0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,  # noqa: E241
       16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,  # noqa: E241,E131,B950
      114,120, 53, 79, 96,109, 72,108, 70, 64, 76, 67,116, 74, 68, 87,  # noqa: E241,E131,B950
      111, 52, 75,119, 49, 34, 82, 81, 95, 65,112, 86,118,110,122,105,  # noqa: E241,E131,B950
       41, 57, 83, 43, 46,102, 40, 89, 38,103, 45, 50, 42,123, 91, 35,  # noqa: E241,E131,B950
      125, 55, 54, 66,124,126, 59, 47, 92, 71,115, 78, 88,107,106, 56,  # noqa: E241,E131,B950
       36,121,117,104,101,100, 69, 73, 99, 63, 94, 93, 39, 37, 61, 48,  # noqa: E241,E131,B950
       58,113, 32, 90, 44, 98, 60, 51, 33, 97, 62, 77, 84, 80, 85,223,  # noqa: E241,E131,B950
      225,216,187,166,229,189,222,188,141,249,148,200,184,136,248,190,  # noqa: E241,E131,B950
      199,170,181,204,138,232,218,183,255,234,220,247,213,203,226,193,  # noqa: E241,E131,B950
      174,172,228,252,217,201,131,230,197,211,145,238,161,179,160,212,  # noqa: E241,E131,B950
      207,221,254,173,202,146,224,151,140,196,205,130,135,133,143,246,  # noqa: E241,E131,B950
      192,159,244,239,185,168,215,144,139,165,180,157,147,186,214,176,  # noqa: E241,E131,B950
      227,231,219,169,175,156,206,198,129,164,150,210,154,177,134,127,  # noqa: E241,E131,B950
      182,128,158,208,162,132,167,209,149,241,153,251,237,236,171,195,  # noqa: E241,E131,B950
      243,233,253,240,194,250,191,155,142,137,245,235,163,242,178,152]  # noqa: E241,E131,B950
    # fmt: on
    for c in password:
        s.append("%c" % scramble_shifts[ord(c)])
    return "".join(s)


def decode_path(path: bytes) -> Tuple[str, str]:
    """Attempt to decode a file path based on encodings known to be used
    in CVS repositories that can be found in the wild.

    Args:
        path: raw bytes path

    Returns:
        A tuple (decoded path, encoding)

    """
    path_encodings = ["ascii", "iso-8859-1", "utf-8"]
    for encoding in path_encodings:
        try:
            how = "ignore" if encoding == path_encodings[-1] else "strict"
            path_str = path.decode(encoding, how)
            break
        except UnicodeError:
            pass
    return path_str, encoding


class CVSProtocolError(Exception):
    pass


class CVSClient:

    # connection to an existing pserver might sometimes fail,
    # retrying the operation usually fixes the issue
    @retry(
        retry=retry_if_exception_type(NotFound),
        stop=stop_after_attempt(max_attempt_number=3),
        reraise=True,
    )
    def connect_pserver(self, hostname, port, username, password):
        if port is None:
            port = CVS_PSERVER_PORT
        if username is None:
            raise NotFound(
                "Username is required for "
                "a pserver connection: %s" % EXAMPLE_PSERVER_URL
            )

        try:
            self.socket = socket.create_connection((hostname, port))
        except ConnectionRefusedError:
            raise NotFound("Could not connect to %s:%s", hostname, port)

        # use empty password if it is None
        scrambled_password = scramble_password(password or "")
        request = "BEGIN AUTH REQUEST\n%s\n%s\n%s\nEND AUTH REQUEST\n" % (
            self.cvsroot_path,
            username,
            scrambled_password,
        )
        print("Request: %s\n" % request)
        self.socket.sendall(request.encode("UTF-8"))

        response = self.conn_read_line()
        if response != b"I LOVE YOU\n":
            raise NotFound(
                "pserver authentication failed for %s:%s: %s"
                % (hostname, port, response)
            )

    def connect_ssh(self, hostname, port, username):
        command = ["ssh"]
        if username is not None:
            # Assume 'auth' contains only a user name.
            # We do not support password authentication with SSH since the
            # anoncvs user is usually granted access without a password.
            command += ["-l", "%s" % username]
        if port is not None:
            command += ["-p", "%d" % port]

        # accept new SSH hosts keys upon first use; changed host keys
        # will require intervention
        command += ["-o", "StrictHostKeyChecking=accept-new"]

        # disable interactive prompting
        command += ["-o", "BatchMode=yes"]

        # disable further option processing by adding '--'
        command += ["--"]

        command += ["%s" % hostname, "cvs", "server"]
        # use non-buffered I/O to match behaviour of self.socket
        self.ssh = subprocess.Popen(
            command, bufsize=0, stdin=subprocess.PIPE, stdout=subprocess.PIPE
        )

    def connect_fake(self):
        command = ["cvs", "server"]
        # use non-buffered I/O to match behaviour of self.socket
        self.ssh = subprocess.Popen(
            command, bufsize=0, stdin=subprocess.PIPE, stdout=subprocess.PIPE
        )

    def conn_read_line(self, require_newline=True):
        if len(self.linebuffer) != 0:
            return self.linebuffer.pop(0)
        buf = b""
        idx = -1
        while idx == -1:
            if len(buf) >= CVS_PROTOCOL_BUFFER_SIZE:
                if require_newline:
                    raise CVSProtocolError(
                        "Overlong response from " "CVS server: %s" % buf
                    )
                else:
                    break
            if self.socket:
                buf += self.socket.recv(CVS_PROTOCOL_BUFFER_SIZE)
            elif self.ssh:
                buf += self.ssh.stdout.read(CVS_PROTOCOL_BUFFER_SIZE)
            else:
                raise Exception("No valid connection")
            if not buf:
                return None
            idx = buf.rfind(b"\n")
        if idx != -1:
            self.linebuffer = buf[: idx + 1].splitlines(keepends=True)
        else:
            if require_newline:
                raise CVSProtocolError("Invalid response from CVS server: %s" % buf)
            else:
                self.linebuffer.append(buf)
        if len(self.incomplete_line) > 0:
            self.linebuffer[0] = self.incomplete_line + self.linebuffer[0]
        if idx != -1:
            self.incomplete_line = buf[idx + 1 :]
        else:
            self.incomplete_line = b""
        return self.linebuffer.pop(0)

    def conn_write(self, data):
        if self.socket:
            return self.socket.sendall(data)
        if self.ssh:
            self.ssh.stdin.write(data)
            return self.ssh.stdin.flush()
        raise Exception("No valid connection")

    def conn_write_str(self, s, encoding="utf-8"):
        return self.conn_write(s.encode(encoding))

    def conn_close(self):
        if self.socket:
            self.socket.close()
        if self.ssh:
            self.ssh.kill()
            try:
                self.ssh.wait(timeout=10)
            except subprocess.TimeoutExpired as e:
                raise subprocess.TimeoutExpired(
                    "Could not terminate " "ssh program: %s" % e
                )

    def __init__(self, url):
        """
        Connect to a CVS server at the specified URL and perform the initial
        CVS protocol handshake.
        """
        self.hostname = url.hostname
        self.cvsroot_path = os.path.dirname(url.path)
        self.cvs_module_name = os.path.basename(url.path)
        self.socket = None
        self.ssh = None
        self.linebuffer = list()
        self.incomplete_line = b""

        if url.scheme == "pserver":
            self.connect_pserver(url.hostname, url.port, url.username, url.password)
        elif url.scheme == "ssh":
            self.connect_ssh(url.hostname, url.port, url.username)
        elif url.scheme == "fake":
            self.connect_fake()
        else:
            raise NotFound("Invalid CVS origin URL '%s'" % url)

        # we should have a connection now
        assert self.socket or self.ssh

        self.conn_write_str(
            "Root %s\nValid-responses %s\nvalid-requests\n"
            "UseUnchanged\n" % (self.cvsroot_path, " ".join(VALID_RESPONSES))
        )
        response = self.conn_read_line()
        if not response:
            raise CVSProtocolError("No response from CVS server")
        try:
            if response[0:15] != b"Valid-requests ":
                raise CVSProtocolError(
                    "Invalid response from " "CVS server: %s" % response
                )
        except IndexError:
            raise CVSProtocolError("Invalid response from CVS server: %s" % response)
        response = self.conn_read_line()
        if response != b"ok\n":
            raise CVSProtocolError("Invalid response from CVS server: %s" % response)

    def __del__(self):
        self.conn_close()

    def _parse_rlog_response(self, fp: IO[bytes]):
        rlog_output = tempfile.TemporaryFile()
        expect_error = False
        for line in fp:
            if expect_error:
                raise CVSProtocolError("CVS server error: %r" % line)
            if line == b"ok\n":
                break
            elif line[0:2] == b"M ":
                rlog_output.write(line[2:])
            elif line[0:8] == b"MT text ":
                rlog_output.write(line[8:-1])
            elif line[0:8] == b"MT date ":
                rlog_output.write(line[8:-1])
            elif line[0:10] == b"MT newline":
                rlog_output.write(line[10:])
            elif line[0:7] == b"error  ":
                expect_error = True
                continue
            else:
                raise CVSProtocolError("Bad CVS protocol response: %r" % line)
        rlog_output.seek(0)
        return rlog_output

    def fetch_rlog(self, path: bytes = b"", state=""):
        if path:
            path_arg, encoding = decode_path(path)
        else:
            path_arg, encoding = self.cvs_module_name, "utf-8"

        if len(state) > 0:
            state_arg = "Argument -s%s\n" % state
        else:
            state_arg = ""
        fp = tempfile.TemporaryFile()
        self.conn_write_str(
            "Global_option -q\n"
            f"{state_arg}"
            "Argument --\n"
            f"Argument {path_arg}\n"
            "rlog\n",
            encoding=encoding,
        )
        while True:
            response = self.conn_read_line()
            if response is None:
                raise CVSProtocolError("No response from CVS server")
            if response[0:2] == b"E ":
                if len(path) > 0 and (
                    response.endswith(b" - ignored\n")
                    or b"could not read RCS file" in response
                ):
                    response = self.conn_read_line()
                    if response not in (b"error  \n", b"ok\n"):
                        raise CVSProtocolError(
                            "Invalid response from CVS server: %s" % response
                        )
                    return None  # requested path does not exist (ignore)
                raise CVSProtocolError("Error response from CVS server: %s" % response)
            fp.write(response)
            if response == b"ok\n":
                break
        fp.seek(0)
        return self._parse_rlog_response(fp)

    def checkout(self, path: bytes, rev: str, dest_dir: bytes, expand_keywords: bool):
        """
        Download a file revision from the cvs server and store
        the file's contents in a temporary file. If expand_keywords is
        set then ask the server to expand RCS keywords in file content.

        From the server's point of view this function behaves much
        like 'cvs update -r rev path'. The server is unaware that
        we do not actually maintain a CVS working copy. Because of
        this it sends more information than we need. We simply skip
        responses that are of no interest to us.
        """
        skip_line = False
        expect_modeline = False
        expect_bytecount = False
        have_bytecount = False
        bytecount = 0

        path_str, encoding = decode_path(path)

        dirname = os.path.dirname(path_str)
        if dirname:
            self.conn_write_str(
                "Directory %s\n%s\n"
                % (dirname, os.path.join(self.cvsroot_path, dirname))
            )
        filename = os.path.basename(path_str)
        co_output = tempfile.NamedTemporaryFile(
            dir=os.fsdecode(dest_dir),
            delete=True,
            prefix="cvsclient-checkout-%s-r%s-" % (filename, rev),
        )
        if expand_keywords:
            # use server-side per-file default expansion rules
            karg = ""
        else:
            # force binary file mode
            karg = "Argument -kb\n"
        # TODO: cvs <= 1.10 servers expect to be given every Directory along the path.
        self.conn_write_str(
            "Directory %s\n%s\n"
            "Global_option -q\n"
            "Argument -r%s\n"
            "%s"
            "Argument --\nArgument %s\nco \n"
            % (
                self.cvs_module_name,
                os.path.join(self.cvsroot_path, self.cvs_module_name),
                rev,
                karg,
                path_str,
            ),
            encoding=encoding,
        )
        while True:
            if have_bytecount:
                if bytecount < 0:
                    raise CVSProtocolError("server sent too much file content data")
                response = self.conn_read_line(require_newline=False)
                if response is None:
                    raise CVSProtocolError("Incomplete response from CVS server")
                if len(response) > bytecount:
                    # When a file lacks a final newline we receive a line which
                    # contains file content as well as CVS protocol response data.
                    # Split last line of file content from CVS protocol data...
                    co_output.write(response[:bytecount])
                    response = response[bytecount:]
                    bytecount = 0
                    # ...and process the CVS protocol response below.
                else:
                    co_output.write(response)
                    bytecount -= len(response)
                    continue
            else:
                response = self.conn_read_line()
            if response[0:2] == b"E ":
                if (
                    b"Skipping `$Log$' keyword due to excessive comment leader"
                    in response
                ):
                    # non fatal error, continue checkout operation without `$Log$'
                    # keyword expansion
                    continue
                raise CVSProtocolError("Error from CVS server: %s" % response)
            if response == b"ok\n":
                if have_bytecount:
                    break
                else:
                    raise CVSProtocolError("server sent 'ok' but no file contents")
            if skip_line:
                skip_line = False
                continue
            elif expect_bytecount:
                try:
                    bytecount = int(response[0:-1])  # strip trailing \n
                except ValueError:
                    raise CVSProtocolError("Bad CVS protocol response: %s" % response)
                have_bytecount = True
                continue
            elif response in (b"M \n", b"MT +updated\n", b"MT -updated\n"):
                continue
            elif response[0:9] == b"MT fname ":
                continue
            elif response.split(b" ")[0] in (
                b"Created",
                b"Checked-in",
                b"Update-existing",
                b"Updated",
                b"Removed",
            ):
                skip_line = True
                continue
            elif response[0:1] == b"/":
                expect_modeline = True
                continue
            elif expect_modeline and response[0:2] == b"u=":
                expect_modeline = False
                expect_bytecount = True
                continue
            elif response[0:2] == b"M ":
                continue
            elif response[0:8] == b"MT text ":
                continue
            elif response[0:10] == b"MT newline":
                continue
            else:
                raise CVSProtocolError("Bad CVS protocol response: %s" % response)
        co_output.seek(0)
        return co_output
